File: samples/vulnerable_app.js
/**
 * Sample vulnerable JavaScript code for ReleaseGuard demo.
 *
 * WARNING: This code contains INTENTIONAL security vulnerabilities.
 * DO NOT use in production.
 */

const express = require('express');
const app = express();

// XSS — unsanitized user input in response
app.get('/search', (req, res) => {
    const query = req.query.q;
    res.send(`<h1>Results for: ${query}</h1>`);
});

// SQL Injection — string concatenation in query
app.get('/user/:id', (req, res) => {
    const query = "SELECT * FROM users WHERE id = '" + req.params.id + "'";
    db.query(query);
});

// Hardcoded credentials
const DB_PASSWORD = "admin123";
const API_SECRET = "sk-prod-very-secret-key-do-not-share";

// Eval with user input
app.post('/calc', (req, res) => {
    const result = eval(req.body.expression);
    res.json({ result });
});

// Command injection
const { exec } = require('child_process');
app.get('/ping', (req, res) => {
    exec('ping -c 1 ' + req.query.host, (err, stdout) => {
        res.send(stdout);
    });
});

// Insecure cookie — no httpOnly, no secure flag
app.get('/login', (req, res) => {
    res.cookie('session', 'abc123', { httpOnly: false, secure: false });
    res.send('logged in');
});

// Open redirect
app.get('/redirect', (req, res) => {
    res.redirect(req.query.url);
});

// Path traversal
const fs = require('fs');
app.get('/file', (req, res) => {
    const content = fs.readFileSync('/uploads/' + req.query.name);
    res.send(content);
});

File: samples/vulnerable_app.py
"""Sample vulnerable Python code for ReleaseGuard demo.

WARNING: This code contains INTENTIONAL security vulnerabilities.
DO NOT use in production.
"""

import hashlib
import os
import subprocess
import sqlite3


# SQL Injection — string formatting in query
def get_user(username):
    conn = sqlite3.connect("app.db")
    cursor = conn.cursor()
    query = f"SELECT * FROM users WHERE username = '{username}'"
    cursor.execute(query)
    return cursor.fetchone()


# Command Injection — unsanitized input in shell command
def run_report(report_name):
    os.system(f"generate_report {report_name}")


# Hardcoded secret
API_KEY = "sk-live-abcdef1234567890abcdef1234567890"
DATABASE_PASSWORD = "super_secret_password_123"


# Weak hashing — MD5 for passwords
def hash_password(password):
    return hashlib.md5(password.encode()).hexdigest()


# Path traversal — user input in file path
def read_file(filename):
    with open(f"/data/uploads/{filename}", "r") as f:
        return f.read()


# Subprocess with shell=True
def ping_host(host):
    result = subprocess.run(f"ping -c 1 {host}", shell=True, capture_output=True)
    return result.stdout


# Insecure deserialization
import pickle

def load_data(data_bytes):
    return pickle.loads(data_bytes)


# SSRF — unvalidated URL
import urllib.request

def fetch_url(url):
    return urllib.request.urlopen(url).read()


# Eval with user input
def calculate(expression):
    return eval(expression)


# Hardcoded JWT secret
JWT_SECRET = "my-super-secret-jwt-key"

def create_token(payload):
    import jwt
    return jwt.encode(payload, JWT_SECRET, algorithm="HS256")
